openedx/core/djangoapps/theming/storage.py

```diff
@@ -2,87 +2,271 @@
+        asset url will be returned.
+
+        Args:
+            name: name of the asset, e.g. 'images/logo.png'
+
+        Returns:
+            url of the asset, e.g. '/static/red-theme/images/logo.png' if current theme is red-theme and logo
+            is provided by red-theme otherwise '/static/images/logo.png'
         """
-        theme_dir = getattr(settings, "COMPREHENSIVE_THEME_DIR", "")
-        if not theme_dir:
-            return None
-        theme_name = os.path.basename(os.path.normpath(theme_dir))
-        return "themes/{name}/".format(name=theme_name)
+        prefix = ''
+        theme_dir = get_current_site_theme_dir()
+
+        # get theme prefix from site address if if asset is accessed via a url
+        if theme_dir:
+            prefix = theme_dir
+
+        # get theme prefix from storage class, if asset is accessed during collectstatic run
+        elif self.prefix:
+            prefix = self.prefix
 
-    def themed(self, name):
+        # join theme prefix with asset name if theme is applied and themed asset exists
+        if prefix and self.themed(name, prefix):
+            name = os.path.join(prefix, name)
+
+        return super(ThemeStorage, self).url(name)
+
+    def themed(self, name, theme):
         """
-        Given a name, return a boolean indicating whether that name exists
-        as a themed asset in the comprehensive theme.
+        Returns True if given asset override is provided by the given theme otherwise returns False.
+        Args:
+            name: asset name e.g. 'images/logo.png'
+            theme: theme name e.g. 'red-theme', 'edx.org'
+
+        Returns:
+            True if given asset override is provided by the given theme otherwise returns False
         """
-        # Nothing can be themed if we don't have a theme location.
-        if not self.theme_location:
-            return False
+        # in debug mode check static asset from within the project directory
+        if settings.DEBUG:
+            themes_location = get_base_theme_dir()
+            # Nothing can be themed if we don't have a theme location or required params.
+            if not all((themes_location, theme, name)):
+                return False
+
+            themed_path = "/".join([
+                themes_location,
+                theme,
+                get_project_root_name(),
+                "static/"
+            ])
+            name = name[1:] if name.startswith("/") else name
+            path = safe_join(themed_path, name)
+            return os.path.exists(path)
+        # in live mode check static asset in the static files dir defined by "STATIC_ROOT" setting
+        else:
+            return self.exists(os.path.join(theme, name))
 
-        path = safe_join(self.theme_location, name)
-        return os.path.exists(path)
 
-    def path(self, name):
+class ComprehensiveThemingCachedFilesMixin(CachedFilesMixin):
+    """
+    Comprehensive theming aware CachedFilesMixin.
```

Comment:
I think we should add a lot more information to this mixin to describe how it works, why we are extending CachedFilesMixin, etc.